packages/lwc-language-server/src/typescript/tsconfig-path-indexer.ts

```diff
@@ -0,0 +1,321 @@
+
+type TSConfigPathItemAttribute = {
+    readonly tsPath: string;
+    readonly filePath: string;
+};
+
+// An internal object representing a path mapping for tsconfig.json file on core
+class TSConfigPathItem {
+    // internal typescript path mapping, e.g., "ui-force-components/modules/force/wireUtils/wireUtils"
+    readonly tsPath: string;
+    // actual file path for the ts file
+    readonly filePath: string;
+
+    constructor(attribute: TSConfigPathItemAttribute) {
+        this.tsPath = attribute.tsPath;
+        this.filePath = attribute.filePath;
+    }
+}
+
+/**
+ * An indexer that stores the TypeScript path mapping info for the Core workspace.
+ *
+ * When using TypeScript for LWCs in the core workspace, the tsconfig.json file's 'paths' attribute needs to be
+ * maintained to ensure that the TypeScript compiler can resolve imported LWC modules. This class serves to maintain
+ * a mapping between LWCs and their paths in tsconfig.json, automatically updating the file when initialized and
+ * whenever a LWC TypeScript file is changed.
+ *
+ * This mapping encompasses all TypeScript LWCs in the core workspace. Each component's full name (namespace/cmpName)
+ * serves as the key, with the corresponding value being an object containing information on how the component should be
+ * mapped in tsconfig.json, thereby enabling TypeScript to locate the component within the file system.
+ */
+export default class TSConfigPathIndexer {
+    readonly coreModulesWithTSConfig: string[];
+    readonly workspaceType: number;
+    // the root path for core directory
+    readonly coreRoot: string;
+    // A map for all TypeScript LWCs on core
+    pathMapping: Map<string, TSConfigPathItem> = new Map<string, TSConfigPathItem>();
+
+    constructor(workspaceRoots: string[]) {
+        this.workspaceType = detectWorkspaceType(workspaceRoots);
+        switch (this.workspaceType) {
+            case WorkspaceType.CORE_ALL:
+                this.coreRoot = workspaceRoots[0];
+                const dirs = fs.readdirSync(this.coreRoot);
+                const subdirs: string[] = [];
+                for (const file of dirs) {
+                    const subdir = path.join(this.coreRoot, file);
+                    if (fs.statSync(subdir).isDirectory()) {
+                        subdirs.push(subdir);
+                    }
+                }
+                this.coreModulesWithTSConfig = this.getCoreModulesWithTSConfig(subdirs);
+                break;
+            case WorkspaceType.CORE_PARTIAL:
+                this.coreRoot = path.join(workspaceRoots[0], '..');
+                this.coreModulesWithTSConfig = this.getCoreModulesWithTSConfig(workspaceRoots);
+                break;
+        }
+    }
+
+    /**
+     * Gets all paths for TypeScript LWC components on core.
+     */
+    get componentEntries(): string[] {
+        const defaultSource = normalize(`${this.coreRoot}/*/modules/*/*/*.ts`);
```

Comment:
Assuming a folder layout like this
```
/ui-force-components
-->/modules
---->/force
------>/one
-------->/one.ts
-------->/util/util.ts
```

`/*/modules/*/*/*.ts` would not include any nested folders(`util.ts`) underneath a component folder. I think that is fine because the importer only can import the main entry file of a module.

Reply:
Yes, that's correct. We only want to include the main component entry ts file and ignore all other utility files.